### PR TITLE
Add support for batch requests to JSON RPC server

### PR DESCRIPTION
**Pull request by RaoulSchaffranek**
Simbolik and kontrol-node exchange many requests, and processing them one request at a time quickly becomes a performance bottleneck. By batching requests together we can avoid the bottleneck. The batching implementation here is part of the JSON RPC spec, see: https://www.jsonrpc.org/specification#batch

### CODE REVIEW

File: pyk/src/pyk/rpc/rpc.py [hunk 1]
```diff
@@ -71,37 +72,105 @@ class JsonRpcMethod(Protocol):
...
-    methods: dict[str, JsonRpcMethod]
+@dataclass(frozen=True)
+class JsonRpcRequest:
 
-    def __init__(self, methods: dict[str, JsonRpcMethod], *args: Any, **kwargs: Any) -> None:
-        self.methods = methods
-        super().__init__(*args, **kwargs)
+    method: str
+    params: Any
+    id: Any
```
Comment (tothtamas28) [inline pyk/src/pyk/rpc/rpc.py]:
```suggestion
    id: int | str
```

File: pyk/src/pyk/rpc/rpc.py [hunk 2]
```diff
@@ -71,37 +72,105 @@ class JsonRpcMethod(Protocol):
     def __call__(self, **kwargs: Any) -> Any: ...
 
 
-class JsonRpcRequestHandler(BaseHTTPRequestHandler):
-    methods: dict[str, JsonRpcMethod]
+@dataclass(frozen=True)
+class JsonRpcRequest:
```
Comment (tothtamas28) [inline pyk/src/pyk/rpc/rpc.py]: Consider just making this a `typing.NamedTuple`:

```suggestion
class JsonRpcRequest(NamedTuple):
```

File: pyk/src/pyk/rpc/rpc.py (hunk 1 above)
Comment (tothtamas28) [reply pyk/src/pyk/rpc/rpc.py]: Also consider making `id` the first attribute.

File: pyk/src/tests/unit/test_json_rpc.py
Comment (tothtamas28) [inline pyk/src/tests/unit/test_json_rpc.py:1]: I think this should rather be an integration test.

File: pyk/src/pyk/rpc/rpc.py [hunk 3]
```diff
@@ -71,37 +72,105 @@ class JsonRpcMethod(Protocol):
...
 
-    def __init__(self, methods: dict[str, JsonRpcMethod], *args: Any, **kwargs: Any) -> None:
-        self.methods = methods
-        super().__init__(*args, **kwargs)
+    method: str
+    params: Any
+    id: Any
+
+    @staticmethod
+    def validate(request_dict: Any, valid_methods: Iterable[str]) -> JsonRpcRequest | JsonRpcError:
```
Comment (tothtamas28) [inline pyk/src/pyk/rpc/rpc.py]: Consider moving this to `JsonRpcRequestHandler`.

File: pyk/src/pyk/rpc/rpc.py [hunk 4]
```diff
@@ -113,44 +182,44 @@ def do_POST(self) -> None:  # noqa: N802
...
-            return
+    def _batch_request(self, requests: list[dict[str, Any]]) -> JsonRpcBatchResult:
+        return JsonRpcBatchResult(tuple(self._single_request(request) for request in requests))
+
+    def _single_request(self, request: dict[str, Any]) -> JsonRpcError | JsonRpcSuccess:
+        validation_result = JsonRpcRequest.validate(request, self.methods.keys())
+        if isinstance(validation_result, JsonRpcError):
+            return validation_result
 
+        method_name = request['method']
```
Comment (tothtamas28) [inline pyk/src/pyk/rpc/rpc.py]:
```suggestion
        method_name = validation_result.method
```

Comment (tothtamas28) [reply pyk/src/pyk/rpc/rpc.py]: But if `JsonRpcRequest` is a `NamedTuple`, you could also just unpack:

```python3
method, args, id = JsonRpcRequest.validate(request, self.methods.keys())
```

File: pyk/src/pyk/rpc/rpc.py (hunk 3 above)
Comment (tothtamas28) [reply pyk/src/pyk/rpc/rpc.py]: Then `valid_methods` can also be removed from the parameters (assuming the method is not static).

**Comment by RaoulSchaffranek** [comment]
Thanks for the review @tothtamas28! Sounds all reasonable to me and I addressed your comments in https://github.com/runtimeverification/k/pull/4688/commits/f8fa54f823fe9202e2a86efd8deb40d77413b872

File: pyk/src/pyk/rpc/rpc.py [hunk 5]
```diff
@@ -71,37 +72,80 @@ class JsonRpcMethod(Protocol):
...
-    def send_json_error(self, code: int, message: str, id: Any = None) -> None:
-        error_dict = {
+class JsonRpcBatchRequest(NamedTuple):
+    requests: tuple[JsonRpcRequest]
+
+
+class JsonRpcResult:
+
+    def encode(self) -> bytes:
+        raise NotImplementedError('Subclasses must implement this method')
```
Comment (tothtamas28) [inline pyk/src/pyk/rpc/rpc.py]:
```suggestion
class JsonRpcResult(ABC):

    @abstractmethod
    def encode(self) -> bytes: ...
```

**Review by tothtamas28 — APPROVED** [review]

**Pull request merged**